dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-fee-record-payment-entity-groups-to-keying-sheet.ts

```diff
@@ -1,65 +1,88 @@
-  exporter: feeRecordEntity.exporter,
-  baseCurrency: feeRecordEntity.baseCurrency,
-  fixedFeeAdjustment: getKeyingSheetAdjustmentForAmount(feeRecordEntity.fixedFeeAdjustment),
-  principalBalanceAdjustment: getKeyingSheetAdjustmentForAmount(feeRecordEntity.principalBalanceAdjustment),
-});
+    if (remainingFeeRecordAmount.eq(0)) {
+      return {
+        ...mapFeeRecordEntityToKeyingSheetRowWithoutFeePayments(feeRecord),
+        feePayments: feePaymentsForRow,
+      };
+    }
 
-const STATUSES_OF_FEE_RECORDS_TO_DISPLAY_ON_KEYING_SHEET: FeeRecordStatus[] = ['READY_TO_KEY', 'RECONCILED'];
+    const lastFeePayment = feePaymentsForRow.at(-1)!;
+    const leftoverFeePaymentAmount = remainingFeeRecordAmount.mul(-1).toNumber();
+    const lastFeePaymentAmountUsed = new Big(lastFeePayment.amount).add(remainingFeeRecordAmount).toNumber();
+
+    lastFeePayment.amount = lastFeePaymentAmountUsed;
+
+    const leftoverFeePayment: KeyingSheetFeePayment = {
+      ...lastFeePayment,
+      amount: leftoverFeePaymentAmount,
+    };
+    allFeePaymentsSortedAscending = orderBy([...allFeePaymentsSortedAscending, leftoverFeePayment], ['amount'], ['asc']);
+
+    return {
+      ...mapFeeRecordEntityToKeyingSheetRowWithoutFeePayments(feeRecord),
+      feePayments: feePaymentsForRow,
+    };
+  });
+};
 
-const getZeroAmountKeyingSheetFeePayment = ({ paymentCurrency }: FeeRecordEntity) => ({
+const getZeroAmountKeyingSheetFeePayment = ({ paymentCurrency }: FeeRecordEntity): KeyingSheetFeePayment => ({
   currency: paymentCurrency,
   amount: 0,
   dateReceived: null,
 });
 
 const mapFeeRecordPaymentEntityGroupToKeyingSheetRows = ({ feeRecords, payments }: FeeRecordPaymentEntityGroup): KeyingSheetRow[] => {
   if (feeRecords.length === 1) {
-    const feePayments =
-      payments.length > 0
-        ? payments.map(({ currency, amount, dateReceived }) => ({ currency, amount, dateReceived }))
-        : [getZeroAmountKeyingSheetFeePayment(feeRecords[0])];
+    const feePayments = payments.length > 0 ? payments.map(mapPaymentEntityToKeyingSheetFeePayment) : [getZeroAmountKeyingSheetFeePayment(feeRecords[0])];
     return [{ ...mapFeeRecordEntityToKeyingSheetRowWithoutFeePayments(feeRecords[0]), feePayments }];
   }
 
-  const { dateReceived } = payments[0];
-  return feeRecords.map((feeRecord) => ({
-    ...mapFeeRecordEntityToKeyingSheetRowWithoutFeePayments(feeRecord),
-    feePayments: [
-      {
-        ...mapFeeRecordEntityToReportedPayments(feeRecord),
-        dateReceived,
-      },
-    ],
-  }));
+  if (payments.length === 1) {
+    const { dateReceived } = payments[0];
+    return feeRecords.map((feeRecord) => ({
+      ...mapFeeRecordEntityToKeyingSheetRowWithoutFeePayments(feeRecord),
+      feePayments: [
+        {
+          ...mapFeeRecordEntityToReportedPayments(feeRecord),
+          dateReceived,
+        },
+      ],
+    }));
+  }
+
+  return getKeyingSheetRowsWithPaymentsAcrossFeeRecords({ feeRecords, payments });
```

Comment:
I think we need to do the splitting once per group when we generate the keying data and then save the split values in the database against the fee records somehow and then read it from there, rather than split as we surface to the ui. We should do this because we want to be certain that the values displayed in the table are what PDC keyed when looking back at the table, so we don't want any uncertainty, the particular case being if we ever decide to change the way we split payments, previously keyed records should still display their split the same.